c/parallel/src/segmented_reduce.cu

```diff
@@ -0,0 +1,481 @@
+    init_t, // 7
+    accum_t); // 8
+}
+
+template <auto* GetPolicy>
+struct dynamic_reduce_policy_t
+{
+  using MaxPolicy = dynamic_reduce_policy_t;
+
+  template <typename F>
+  cudaError_t Invoke(int device_ptx_version, F& op)
+  {
+    return op.template Invoke<segmented_reduce_runtime_tuning_policy>(GetPolicy(device_ptx_version, accumulator_type));
+  }
+
+  cccl_type_info accumulator_type;
+};
+
+struct segmented_reduce_kernel_source
+{
+  cccl_device_segmented_reduce_build_result_t& build;
+
+  std::size_t AccumSize() const
+  {
+    return build.accumulator_size;
+  }
+  CUkernel SegmentedReduceKernel() const
+  {
+    return build.segmented_reduce_kernel;
+  }
+};
+
+} // namespace segmented_reduce
+
+extern "C" CCCL_C_API CUresult cccl_device_segmented_reduce_build(
+  cccl_device_segmented_reduce_build_result_t* build,
+  cccl_iterator_t input_it,
+  cccl_iterator_t output_it,
+  cccl_iterator_t start_offset_it,
+  cccl_iterator_t end_offset_it,
+  cccl_op_t op,
+  cccl_value_t init,
+  int cc_major,
+  int cc_minor,
+  const char* cub_path,
+  const char* thrust_path,
+  const char* libcudacxx_path,
+  const char* ctk_path) noexcept
+{
+  CUresult error = CUDA_SUCCESS;
+
+  try
+  {
+    const char* name = "device_segmented_reduce";
+
+    const int cc                       = cc_major * 10 + cc_minor;
+    const cccl_type_info accum_t       = segmented_reduce::get_accumulator_type(op, input_it, init);
+    const auto policy                  = segmented_reduce::get_policy(cc, accum_t);
+    const auto accum_cpp               = cccl_type_enum_to_string(accum_t.type);
+    const auto input_it_value_t        = cccl_type_enum_to_string(input_it.value_type.type);
+    const auto start_offset_it_value_t = cccl_type_enum_to_string(start_offset_it.value_type.type);
+    const auto end_offset_it_value_t   = cccl_type_enum_to_string(end_offset_it.value_type.type);
+    // OffsetT is checked to match have 64-bit size
+    const auto offset_t = cccl_type_enum_to_string(cccl_type_enum::UINT64);
+
+    const std::string input_iterator_src =
+      make_kernel_input_iterator(offset_t, "input_iterator_t", input_it_value_t, input_it);
+    const std::string output_iterator_src =
+      make_kernel_output_iterator(offset_t, "output_iterator_t", accum_cpp, output_it);
+
+    const std::string start_offset_iterator_src =
+      make_kernel_input_iterator(offset_t, "start_offset_iterator_t", start_offset_it_value_t, start_offset_it);
+    const std::string end_offset_iterator_src =
+      make_kernel_input_iterator(offset_t, "end_offset_iterator_t", end_offset_it_value_t, end_offset_it);
+
+    const std::string op_src = make_kernel_user_binary_operator(accum_cpp, op);
+
+    // agent_policy_t is to specify parameters like policy_hub does in dispatch_reduce.cuh
+    constexpr std::string_view program_preamble_template =
+      "#include <cub/block/block_reduce.cuh>\n"
+      "#include <cub/device/dispatch/kernels/segmented_reduce.cuh>\n"
+      "\n"
```

Comment:
Q: Why can't this be a raw string literal?

Reply:
I was following along with what is done in "reduce.cu" and "scan.cu".

Will change to use raw string literal here and in "test_utils.h" as well as in other files. Going to be easier to read and maintain.

Reply:
Existing code is not always the best code :) Please use a raw string literal. Thx!